### PR TITLE
Add OpenMetrics support

### LINKED ISSUE #536: Prometheus support
Hey @jhaals,

first of all, thanks so much for yopass, we're big fan here at SoundCloud! It's our favorite tool to quickly share some credentials with others.

We wonder if you were to accept a submission to add Prometheus monitoring to yopass, to track number of requests and errors? I saw that the prometheus client library is already in the list of dependencies. So we'd only need to load the library, add one metric, and expose the `/metrics` handler. https://godoc.org/github.com/prometheus/client_golang/prometheus/promhttp#InstrumentHandlerDuration

Cheers!

### PR DESCRIPTION
**Pull request by grobie**
Instrument Yopass' HTTP handler and record served HTTP requests by
method, path and reponse code, and request duration by method and path.

Metrics are optionally served in the [OpenMetrics][] text format on a
separate HTTP server under /metrics. The metrics server can be enabled
by setting the --metrics-port flag. For now listen IP address and TLS
settings are used for both HTTP server instances.

[OpenMetrics]: https://openmetrics.io/

Resolves #536. Depends on #537. 

<details>
  <summary>Example /metrics output:</summary>

```
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 0
go_gc_duration_seconds{quantile="0.25"} 0
go_gc_duration_seconds{quantile="0.5"} 0
go_gc_duration_seconds{quantile="0.75"} 0
go_gc_duration_seconds{quantile="1"} 0
go_gc_duration_seconds_sum 0
go_gc_duration_seconds_count 0
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 7
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.14.6"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 2.214048e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.214048e+06
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.444929e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 1350
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 0
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 3.436808e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 2.214048e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 6.3340544e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 3.211264e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 9672
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 6.3340544e+07
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 6.6551808e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 0
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 11022
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 13888
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 63784
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 65536
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 4.473924e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.278903e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 557056
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 557056
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 7.3351424e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 10
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.05
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 8192
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 79
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 1.3897728e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.59674120894e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.340944384e+09
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
# HELP yopass_request_duration_seconds Histogram of HTTP request latencies by method and path.
# TYPE yopass_request_duration_seconds histogram
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.005"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.01"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.025"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.05"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.1"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.25"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="0.5"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="1"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="2.5"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="5"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="10"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/",le="+Inf"} 1
yopass_request_duration_seconds_sum{method="GET",path="/"} 0.000121674
yopass_request_duration_seconds_count{method="GET",path="/"} 1
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.005"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.01"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.025"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.05"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.1"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.25"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="0.5"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="1"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="2.5"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="5"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="10"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/metrics",le="+Inf"} 5
yopass_request_duration_seconds_sum{method="GET",path="/metrics"} 0.012709560000000002
yopass_request_duration_seconds_count{method="GET",path="/metrics"} 5
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.005"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.01"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.025"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.05"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.1"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.25"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="0.5"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="1"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="2.5"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="5"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="10"} 11
yopass_request_duration_seconds_bucket{method="GET",path="/secret/:key",le="+Inf"} 11
yopass_request_duration_seconds_sum{method="GET",path="/secret/:key"} 0.007584642
yopass_request_duration_seconds_count{method="GET",path="/secret/:key"} 11
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.005"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.01"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.025"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.05"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.1"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.25"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="0.5"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="1"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="2.5"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="5"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="10"} 2
yopass_request_duration_seconds_bucket{method="POST",path="/secret",le="+Inf"} 2
yopass_request_duration_seconds_sum{method="POST",path="/secret"} 0.002809964
yopass_request_duration_seconds_count{method="POST",path="/secret"} 2
# HELP yopass_requests_total Total number of requests served by HTTP method, path and response code.
# TYPE yopass_requests_total counter
yopass_requests_total{code="200",method="GET",path="/metrics"} 5
yopass_requests_total{code="200",method="GET",path="/secret/:key"} 10
yopass_requests_total{code="200",method="POST",path="/secret"} 1
yopass_requests_total{code="400",method="POST",path="/secret"} 1
yopass_requests_total{code="404",method="GET",path="/"} 1
yopass_requests_total{code="404",method="GET",path="/secret/:key"} 1
```

</details>

### CODE REVIEW

**Comment by grobie** [comment]
Hej @jhaals. This is ready for review, let me know what you'd like to see changed / where you'd prefer a different style or such. Happy to make any changes.

**Comment by jhaals** [comment]
Thanks a lot for the PR @grobie, I will happily accept these changes. Would you also be able to add some notes about metrics in the readme?

**Comment by grobie** [comment]
@jhaals sure thing. Like that? https://github.com/jhaals/yopass/blob/b7415f00b90217b9465ca4db3096145049fecf84/README.md#monitoring

**Comment by jhaals** [comment]
Fantastic 👍

File: cmd/yopass/main.go [hunk 1]
```diff
@@ -18,6 +20,7 @@ var (
 	maxLength = pflag.Int("max-length", 10000, "max length of encrypted secret")
 	memcached = pflag.String("memcached", "localhost:11211", "Memcached address")
 	port      = pflag.Int("port", 1337, "listen port")
+	_         = pflag.Int("metrics-port", -1, "metrics server listen port")
```
Comment (jhaals) [inline cmd/yopass/main.go:23]: is this supposed to be underscore?

Comment (grobie) [reply cmd/yopass/main.go:23]: Hehe. So I realized that all these global variables are unused, as you use viper to access the flag values. I'll change it for consistency.

Comment (grobie) [reply cmd/yopass/main.go]: Done. It might make sense to remove all these variables at some point though to make it clear that they aren't used.

Comment (jhaals) [reply cmd/yopass/main.go]: @grobie I did not know that the variables where unused, it must have been a leftover from the time when I used `flag`. It make sense to clean them up in a separate PR. I'm going to merge this one. Thanks a lot for adding metrics to yopass ❤️

**Pull request merged**